android/envoy/src/main/kotlin/org/greatfire/envoy/EnvoyTest.kt

```diff
@@ -89,196 +100,213 @@ data class EnvoyTest(
+        if (proxyUrl != null && !proxyIsEnvoy) {
+            proxyUrl = proxyUrl!!
         }
+
+        val cronetEngine = CronetNetworking.buildEngine(
+            context = context,
+            cacheFolder = null, // no cache XXX?
+            proxyUrl = proxyUrl,
+            resolverRules = resolverRules,
+            cacheSize = 0, // what are the units here?
+        )
+        val callback = TestUrlRequestCallback()
+        // We're just making a standard request via Cronet to a standard proxy
+        val requestBuilder = cronetEngine.newUrlRequestBuilder(
+            testUrl,
+            callback,
+            Executors.newCachedThreadPool()
+        )
+
+        val request = requestBuilder.build()
+        request.start()
+        callback.blockForResponse()
+
+        // test is now complete
+        return callback.succeeded
     }
 
-    fun stopService() {
-        // stop the associated service
-        // this is called to stop unused services
-        when (testType) {
-            EnvoyServiceType.HYSTERIA2 -> state.iep?.let { it.stop(IEnvoyProxy.Hysteria2) }
-            // EnvoyServiceType.SHADOWSOCKS -> shadowsocks?.let { it.stop() }
-            // EnvoyServiceType.SHADOWSOCKS -> state.ctx!!.stopService(shadowsocksIntent)
-            EnvoyServiceType.V2SRTP -> state.iep?.let { it.stop(IEnvoyProxy.V2RaySrtp) }
-            EnvoyServiceType.V2WECHAT -> state.iep?.let { it.stop(IEnvoyProxy.V2RayWechat) }
-            EnvoyServiceType.CRONET_MASQUE,
-            EnvoyServiceType.OKHTTP_MASQUE -> {
-                // this is currently a no-op, but call it in case it ever isn't :)
-                state.iep?.let { it.stop(IEnvoyProxy.Masque) }
-            }
-            EnvoyServiceType.CRONET_ENVOY,
-            EnvoyServiceType.OKHTTP_ENVOY,
-            EnvoyServiceType.CRONET_PROXY,
-            EnvoyServiceType.OKHTTP_PROXY -> {
-                // no service for these
-                Log.d(TAG, "No service to stop for $testType")
-            }
-            EnvoyServiceType.HTTP_ECH -> {
-                Log.d(TAG, "TODO: we can stop the ECH proxy")
-            }
-            EnvoyServiceType.DIRECT -> "no op"
-            else -> {
-                Log.e(TAG, "Tried to stop an unknown service $testType")
-            }
+    fun getProxy(
+        proxyType: Proxy.Type,
+        host: String,
+        port: Int,
+    ): Proxy {
+        val addr = InetSocketAddress(host, port)
+        return Proxy(proxyType, addr)
+    }
+
+    // was getEnvoyUrl but returns only ech url? (if available)
+    fun getEchUrl(): String {
+        // XXX this needs cleanup? get the Envoy URL from IEP
+        state.iep?.let {
+            return it.echProxyUrl
         }
+        Log.e(TAG, "No EchProxyUrl in IEP")
+        return ""
     }
 
     fun startTimer() {
-        getTimer() // this starts the timer as a side effect
+        checkTimer() // this starts the timer as a side effect
```

Comment:
should checkTimer() be startTimer()? This seems like cruft from me refactoring your timing code out of the test code... I wasn't really sure how you wanted it, since the timer is only used by the reporting, and thought you might clean it up at some point... I think the code used to have a startTimer(), but making it it's own class, the timer now starts implicitly when you make an instance. checkTimer is also kind of a weird name for a function that returns a timer. 

You don't necessarily need to address it here, but it's probably easy to clean up a bit how the timer is used. I was just trying to clean things up a little...

Reply:
see above, i just needed to rename it something other than "get".  "start" is fine too, i'll check how it's being used while i'm at it.

Reply:
it looked like getTimer/checkTimer was only being used within these methods to access the timer that's in the same class anyway.  i removed checkTimer and changed startTimer to create a timer if none exists, and stopTimer/timeSpent now return 0 if no timer exists.